src/FsAutoComplete.Core/Sourcelink.fs

```diff
@@ -0,0 +1,169 @@
+module FsAutoComplete.Sourcelink
+
+open System.IO
+open System.Reflection.Metadata
+open System.Reflection.PortableExecutable
+open System.Text.RegularExpressions
+open Newtonsoft.Json
+open FSharp.Data
+open FsAutoComplete.Logging
+
+let logger = LogProvider.getLoggerByName "FsAutoComplete.Sourcelink"
+
+let private sourceLinkGuid = System.Guid "CC110556-A091-4D38-9FEC-25AB9A351A6A"
+let private embeddedSourceGuid = System.Guid "0E8A571B-6926-466E-B4AD-8AB04611F5FE"
+
+let private toHex (bytes: byte[]) =
+    System.BitConverter.ToString(bytes).Replace("-", "").ToLowerInvariant()
+
+type SourceLinkJson =
+ { documents: System.Collections.Generic.Dictionary<string,string> }
+
+type private Document =
+    { Name: string
+      Hash: byte[]
+      Language: System.Guid
+      IsEmbedded: bool }
+
+let private pdbForDll (dllPath: string) =
+    Path.ChangeExtension(dllPath, ".pdb")
+
+let private tryGetSourcesForPdb (pdbPath: string) =
+    logger.info (Log.setMessage "Reading metadata information for PDB {pdbPath}" >> Log.addContextDestructured "pdbPath" pdbPath)
+    match File.Exists pdbPath with
+    | true ->
+        let pdbData = File.OpenRead pdbPath
+        MetadataReaderProvider.FromPortablePdbStream pdbData |> Some
+    | false ->
+        None
+
+let private tryGetSourcesForDll (dllPath: string) =
+    logger.info (Log.setMessage "Reading metadata information for DLL {dllPath}" >> Log.addContextDestructured "dllPath" dllPath)
+    let file = File.OpenRead dllPath
+    let embeddedReader = new PEReader(file)
```

Comment:
How does the stream ownership work with regards to the `PEReader`? Like does line 153 handle disposing this correctly, or do we need to dispose the `PEReader` here?

Reply:
You've got it right. Ownership of the stream is transferred using this constructor, and when the PEReader is disposed is closes the underlying stream.